Separating QueryParseContext and QueryShardContext

cbuescher opened the pull request:
The parse() method we are currently splitting into a query parsing part (fromXContent()) and a lucene query generating part (toQuery()) in the feature refactoring branch (#10217) takes QueryParseContext as an argument. At some point we need to excute these two steps in different phases on the coordinating node and the shards. This protoype PR tries to anticipate this by introducing a new QueryCreationContext object as the argument of the query-generating methods (toQuery() and all dependent methods). 
As a first step the existing QueryParseContext is renamed to QueryCreationContext, which itself acts as a wrapper around a new lightweight QueryParseContext that should only contain the functionality used during query parsing. In subsequent steps we can then try to separate the different behaviour needed in the two context objects.

This is WIP, the separation of the two context is not completely clean and local variables are not renames. Also the naming of the new context object is completely open. Would like to open this PR to get feedback and discuss. 

PR goes against query refactoring branch.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BaseQueryParserTemp.java:
```diff
@@ -32,7 +32,7 @@
 
     @Override
-    public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOException, QueryParsingException {
+    public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOException, QueryCreationException {
```
I think the exception should still be QueryParsingException here

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -126,10 +126,10 @@ public ParsedQuery parse(QueryBuilder queryBuilder) {
             parser = XContentFactory.xContent(bytes).createParser(bytes);
             return parse(cache.get(), parser);
-        } catch (QueryParsingException e) {
+        } catch (QueryCreationException e) {
```
I am conflicted on whether we should keep query parsing exception here. Anyways this method will go away right? Can we mark //norelease all the methods that we expect to be able to remove once we are done with the refactoring?

cbuescher replied:
This particular method seems only to be used in tests at the moment. I'm unsure about what all the parse() methods in IndexQueryParserService are doing since I haven't directly used them yet. Do you think those will be calles as part of the query generation or are they also called in fromXContent? Need to investigate this.

javanna replied:
I had a look at all these parse methods. We might need to clean them up, we have too many variants of it, most of them are used in tests. But in general they are used by any component that needs to parse queries: percolator, highlighting (supports a separate highlight_query that gets parsed as part of highlighting), query rescorer, translog (old delete_by_query)...... I think those parse methods should be converted to fromXContent and return a QueryBuilder instead.

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -225,26 +227,26 @@ public ParsedQuery parseInnerFilter(XContentParser parser) throws IOException {
     }
 
-    public QueryParseContext getParseContext() {
+    public QueryCreationContext getParseContext() {
```
maybe we should rename the method here as well?

javanna commented on core/src/main/java/org/elasticsearch/index/percolator/PercolatorQueriesRegistry.java:
```diff
@@ -208,12 +208,12 @@ private Query parseQuery(String type, XContentParser parser) {
             return queryParserService.parseInnerQuery(context);
         } catch (IOException e) {
-            throw new QueryParsingException(context, "Failed to parse", e);
+            throw new QueryCreationException(context, "Failed to parse", e);
```
seems weird to have a query creation exception with text "failed to parse". This is still a method that parse all the way from xcontent to lucene query, not 100% sure what we should be doing here. Maybe the change is ok but we need to add some TODO or //norelease that says we have to get back to this method and split it in two phases at some point.

cbuescher replied:
Haven't realized this so far, this also means that we will not be able to remove the parser part from whatever context object we use here since we need both functionalitites still. Any feeling about how this might eventually change after the refactoring? Will we pass builders to this method then?

javanna replied:
this will need to be refactored along the same lines as our refactoring, as mentioned in another comment

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+
+    private final Version indexVersionCreated;
+
+    private final IndexQueryParserService indexQueryParser;
```
I wonder if this is needed only temporarily here, otherwise we might need to think about renaming it cause it might not be about parsing anymore?

cbuescher replied:
`IndexQueryParserService` is used internally all over the place to get all kinds of services, however I'm not sure if we will also need it on the parsing side and we should rename it.

cbuescher replied:
Having worked a bit more on moving parsing methods into new the QueryParseContext, I tend to think we should rename it here into something like IndexQueryService. However, the problem currently is that `IndexQueryParserService` contains a bunch of `Query parse...(QueryParseContext ...)` methods that e.g. parse plain byte[] or ByteReference queries and return lucene queries, those need to be refactored as well I think.

javanna replied:
yes they need to be refactored, I think they should be renamed to fromXContent, I can agree with the rename of the service idea.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+
+//    private XContentParser parser;
+
+    private ParseFieldMatcher parseFieldMatcher;
```
this one should only be about parsing, so it should not be needed here?

cbuescher replied:
We seem to use it currently in three toQuery() methods when resolving the rewrite method. In PrefixQB, RexexpQB, WildcardQB there is a call to context.parseFieldMatch() which is used to resolve the various rewrite constants.

javanna replied:
that's fine then, I double checked, naming is a bit confusing but it's ok at least for now

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+
+    private final MapperQueryParser queryParser = new MapperQueryParser(this);
+
+//    private XContentParser parser;
```
remove?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+//     */
+//    public boolean isFilter() {
+//        return isFilter;
+//    }
```
don't we need this?

cbuescher replied:
I'm unsure, was hoping we only need this in fromXContent(), thats why I removed it from this context. However, we might need something similar so this when traversing the query builder tree when generating the lucene query. Something on my todo list to investigate.

cbuescher replied:
I think we will need this on the shard, which is a huge pita since we basically have to go through all the `doToQuery()` methods so far and wrap calls to `toQuery() like we now do in `QueryParseContext#parseInnerFilter()`

javanna replied:
I suspected this would be the case, should we make this change outside of this PR? seems like this already needs fixing no? Then we would simply need to move isFilter to QueryCreationContext, and mark //norelease the one in QueryParseContext which will go away once we refactored all the queries?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+     */
+    @Nullable
+    @Deprecated
+    public Query parseInnerFilter(String queryName) throws IOException, QueryCreationException {
```
add //norelease here too?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+     */
+    @Nullable
+    @Deprecated
+    public Query parseInnerFilter() throws QueryCreationException, IOException {
```
add //norelease here?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+        return indexVersionCreated;
+    }
+
+    public QueryParseContext parseContext() {
```
I thought the QueryParseContext needs to expose the QueryCreationContext for the queries that are not refactored yet. Then the QueryCreationContext shouldn't need to expose the QueryParseContext? Why do we need this?

cbuescher replied:
Good thought, tried to remove this direction of dependency, this opened a whole lot of new issues I'll try to resolve before I rebase here.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class QueryCreationContext {
```
naming... I think this is too generic, we mean that this context allows to create a lucene query, it will become even more confusing once we remove the "Builder" suffix from all the query builders. Maybe QueryShardContext or something along those lines, trying to mention the distinction between coord node and data node (shards)?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,357 @@
+public class QueryCreationContext {
+
+    private static final ParseField CACHE = new ParseField("_cache").withAllDeprecated("Elasticsearch makes its own caching decisions");
+    private static final ParseField CACHE_KEY = new ParseField("_cache_key").withAllDeprecated("Filters are always used as cache keys");
```
do we need these two constants here? I thought we shouldn't need them

cbuescher replied:
Yes, I think I can remove them.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,76 @@
 
-    public void parser(XContentParser parser) {
-        this.parser = parser;
+    public QueryCreationContext context() {
```
rename the method accordingly to the name given to the QueryCreationContext?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -418,19 +201,12 @@ public long nowInMillis() {
-    public Version indexVersionCreated() {
-        return indexVersionCreated;
+    //norelease remove dependency on context here
+    public MappedFieldType fieldMapper(String name) {
```
since we have a getter for the context, can't we simply remove and replace callers code with queryParseContext.context().fieldMapper(name) ?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,100 +161,36 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
-                return fieldMapping;
-            }
-        }
+    public String defaultField() {
```
I think this should be moved to QueryCreationContext and exposed here through it temporarily

cbuescher replied:
will do.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,100 +161,36 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
 
-    public void setMapUnmappedFieldAsString(boolean mapUnmappedFieldAsString) {
-        this.mapUnmappedFieldAsString = mapUnmappedFieldAsString;
+    public Similarity searchSimilarity() {
```
I think this should be moved to QueryCreationContext and exposed here through it temporarily

cbuescher replied:
will do.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,100 +161,36 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
-
-    public ObjectMapper getObjectMapper(String name) {
-        return indexQueryParser.mapperService.getObjectMapper(name, getTypes());
+    public boolean isFilter() {
```
I have the feeling this should not be handled in parse phase but only in toQuery....

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,100 +161,36 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
-            return fieldType.searchQuoteAnalyzer();
-        }
-        return mapperService().searchQuoteAnalyzer();
+    public AnalysisService analysisService() {
```
I think this should be moved to QueryCreationContext and exposed here through it temporarily

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,76 @@
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseFieldMatcher = parseFieldMatcher;
-    }
+    private final QueryCreationContext context;
```
rename the variable?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,76 @@
 
-    public ParseFieldMatcher parseFieldMatcher() {
-        return parseFieldMatcher;
+    public QueryParseContext(QueryCreationContext context) {
```
rename the argument?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParsingException.java:
```diff
@@ -62,7 +59,7 @@ public QueryParsingException(QueryParseContext parseContext, String msg, Throwab
     /**
      * This constructor is provided for use in unit tests where a
-     * {@link QueryParseContext} may not be available
+     * {@link QueryCreationContext} may not be available
```
is this rename ok?

cbuescher replied:
No, Eclipse was too eager here.

javanna commented on core/src/main/java/org/elasticsearch/index/query/TemplateQueryParser.java:
```diff
@@ -80,9 +80,9 @@ public Query parse(QueryParseContext parseContext) throws IOException {
+            final QueryCreationContext context = new QueryCreationContext(parseContext.context().index(), parseContext.context().indexQueryParserService());
             context.reset(qSourceParser);
-            return context.parseInnerQuery();
+            return context.parseContext().parseInnerQuery();
```
not sure why we go through the creation of QueryCreationContext to then retrieve the QueryParseContext, when we have the QueryParseContext in the first place

javanna commented on core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java:
```diff
@@ -56,11 +56,11 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
         byte[] querySource = parser.binaryValue();
         try (XContentParser qSourceParser = XContentFactory.xContent(querySource).createParser(querySource)) {
-            final QueryParseContext context = new QueryParseContext(parseContext.index(), parseContext.indexQueryParserService());
+            final QueryCreationContext context = new QueryCreationContext(parseContext.index(), parseContext.indexQueryParserService());
```
similar to the comment above, just use the QueryParseContext that we have here? We will deal with what needs to be done when we refactor this query?

javanna commented on core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryParser.java:
```diff
@@ -221,7 +221,7 @@ private String parseFiltersAndFunctions(QueryParseContext parseContext, XContent
                     } else {
                         if ("filter".equals(currentFieldName)) {
-                            filter = parseContext.parseInnerFilter();
+                            filter = parseContext.context().parseInnerFilter();
```
I am not sure why we moved these parseInnerFilter methods to QueryCreationContext. Didn't they make more sense in QueryParseContext (although they will be removed at some point)?

javanna commented on core/src/main/java/org/elasticsearch/index/query/support/NestedInnerQueryParseSupport.java:
```diff
@@ -41,7 +40,7 @@
 public class NestedInnerQueryParseSupport {
 
-    protected final QueryParseContext parseContext;
+    protected final QueryCreationContext creationContext;
```
I am not sure whether this change should be postponed to the refactoring of nested queries, what do you think?

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -305,12 +310,21 @@ public void testSerialization() throws IOException {
     /**
+     * @return a new {@link QueryCreationContext} based on the base test index and queryParserService
+     */
+    protected static QueryCreationContext createContext() {
```
improve the method name?

javanna commented:
left a bunch of comments

cbuescher commented:
I continued thinning down the new lightweight QueryParseContext, addressing some of the comments and trying to make QueryParseContext only refer back to its inner QueryCreationContext, not the other way round. However, I'm thinking if we shouldn't reverse this, making the new QueryParseContext that should only contain the parser, some additional state and the helper methods (parseInnerQuerBuilder etc...) an inner part of the new QueryCreationContext, exposing it there through "getParseContext()" in all the places (essentially the old `parse(QueryParseContext)` methods that haven't been refactored yet. So we could use the new `QueryParseContext` only in methods that we sure that only need parsing (like the parsers `fromXContent()`) and making the un-refactored `parse()` methods get a QueryCreationContext where they can retrieve a parse context if need be.

javanna commented on core/src/main/java/org/elasticsearch/action/admin/indices/validate/query/TransportValidateQueryAction.java:
```diff
@@ -189,7 +190,10 @@ protected ShardValidateQueryResponse shardOperation(ShardValidateQueryRequest re
+        } catch (QueryCreationException e) {
+            valid = false;
+            error = e.getDetailedMessage();
         } catch (QueryParsingException e) {
```
I think this catch is unreachable now?

javanna commented on core/src/main/java/org/elasticsearch/index/percolator/PercolatorQueriesRegistry.java:
```diff
@@ -184,14 +185,15 @@ Query parsePercolatorDocument(String id, BytesReference source) {
     }
 
+    //norelease this method parses from xcontent to lucene query, need to re-investigate how to split context here
     private Query parseQuery(String type, XContentParser parser) {
```
I think we will need to do the same that we are doing for searches, percolator queries will be parsed to our own query format when indexed, at execution time they will be converted to lucene query. This will come with a little cost, given that we currently parse directly to lucene, but it will also solve problems. Have a look at #12179 , maybe we can mention it here in the comment.

cbuescher replied:
So do you think we need to do the parsing to QueryBuilder already in this PR or should we do it later?

javanna replied:
no I would do it later, definitely

javanna replied:
clarify the comment and add reference to the open issue? We don't need to investigate, we just have to do it when ready :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/CommonTermsQueryParser.java:
```diff
@@ -87,7 +87,8 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                         analyzer = parser.text();
-                        if (parseContext.analysisService().analyzer(analyzer) == null) {
+                        //norelease we have to delay this check to the shard when parsing on the coordinating node
+                        if (parseContext.context().analysisService().analyzer(analyzer) == null) {
```
maybe we should change this and move the check to the data node already? not sure //norelease is good here, or maybe I am missing something

cbuescher replied:
Yes, I did that with my last WIP (not pushed yet)

javanna commented on core/src/main/java/org/elasticsearch/index/query/MatchQueryParser.java:
```diff
@@ -94,7 +94,8 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                     } else if ("analyzer".equals(currentFieldName)) {
                         String analyzer = parser.text();
-                        if (parseContext.analysisService().analyzer(analyzer) == null) {
+                        //norelease we have to delay this check to the shard when parsing on the coordinating node
```
I think this is a todo for when we refactor this query right?

cbuescher replied:
yes, should I remove the comment?

javanna replied:
yes I think we can remove all of these comments for queries that are not refactored yet. And we shoud address now similar comments for queries that are already refactored.

javanna commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java:
```diff
@@ -95,7 +95,7 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
 
         MoreLikeThisQuery mltQuery = new MoreLikeThisQuery();
-        mltQuery.setSimilarity(parseContext.searchSimilarity());
+        mltQuery.setSimilarity(parseContext.context().searchSimilarity());
```
this should maybe be moved to the data node as well once we refactor this query?

javanna commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java:
```diff
@@ -142,7 +142,7 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                     mltQuery.setMinimumShouldMatch(parser.text());
                 } else if ("analyzer".equals(currentFieldName)) {
-                    analyzer = parseContext.analysisService().analyzer(parser.text());
+                    analyzer = parseContext.context().analysisService().analyzer(parser.text());
```
this as well will need to be moved to the data node once we refactor this query?

javanna commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java:
```diff
@@ -165,7 +165,7 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                     while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
                         String field = parser.text();
-                        MappedFieldType fieldType = parseContext.fieldMapper(field);
+                        MappedFieldType fieldType = parseContext.context().fieldMapper(field);
```
and this as well?

javanna commented on core/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryParser.java:
```diff
@@ -214,14 +214,14 @@ else if (parseContext.parseFieldMatcher().match(currentFieldName, Fields.UNLIKE)
         boolean useDefaultField = (moreLikeFields == null);
         if (useDefaultField) {
-            moreLikeFields = Lists.newArrayList(parseContext.defaultField());
+            moreLikeFields = Lists.newArrayList(parseContext.context().defaultField());
```
and this too?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationContext.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    /** Gets the search quote nalyzer for the given field, or the default if there is none present for the field
+     * TODO: remove this by moving defaults into mappers themselves
```
do you want to address these two TODOs in this PR or as a separate change?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryCreationException.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class QueryCreationException extends ElasticsearchException {
+
+    static final int UNKNOWN_POSITION = -1;
```
we can drop this no?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,103 @@
     private XContentParser parser;
-
+    private final Index index;
+    private boolean isFilter;
```
shall we mark the isFilter stuff //norelease so we don't forget to remove it?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,103 @@
-
-    public void parser(XContentParser parser) {
-        this.parser = parser;
+    public QueryCreationContext context() {
```
rename to `getQueryCreationContext()`

javanna replied:
Also, will this stay around or go away at some point? If it's supposed to be gone we should mark as //norelease

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,103 @@
+     */
+    @Nullable
+    @Deprecated
+    public Query parseInnerQuery() throws IOException, QueryCreationException {
```
should it be marked //norelease too?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,123 +188,26 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
 
     /**
-     * Return whether the setting is deprecated.
+     * @return a new ParseContext object has independent parser state
```
missing "that" in the javadocs?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,123 +188,26 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
-    public Version indexVersionCreated() {
-        return indexVersionCreated;
+    public QueryParseContext copy() {
+        return new QueryParseContext(this.context);
```
before we were creating a new context when needed, now we call this method which does the same. Was wondering what this new method buys us to be honest, maybe we can do without?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryStringQueryParser.java:
```diff
@@ -214,16 +214,16 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        qpSettings.defaultAnalyzer(parseContext.mapperService().searchAnalyzer());
-        qpSettings.defaultQuoteAnalyzer(parseContext.mapperService().searchQuoteAnalyzer());
+        qpSettings.defaultAnalyzer(parseContext.context().mapperService().searchAnalyzer());
+        qpSettings.defaultQuoteAnalyzer(parseContext.context().mapperService().searchQuoteAnalyzer());
```
lots of things here will need to be moved to the data node too once we refactor this query I think

javanna commented on core/src/main/java/org/elasticsearch/index/query/SimpleQueryStringParser.java:
```diff
@@ -109,12 +109,13 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                             fField = parser.text();
                         }
 
+                        //norelease we need to delay access to the mapper service to the shard
```
I think we should fix this now given that this query is already refactored, this seems like a bug

cbuescher replied:
Yes, moved this to `toQuery`

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -221,7 +226,7 @@ public void testFromXContent() throws IOException {
     @Test
     public void testToQuery() throws IOException {
-        QueryParseContext context = createContext();
+        QueryCreationContext context = createContext();
```
I would rename this createContext method too once we are settled on the naming for QueryCreationContext

javanna commented on core/src/test/java/org/elasticsearch/index/query/SimpleIndexQueryParserTests.java:
```diff
@@ -2361,6 +2425,7 @@ public void testProperErrorMessagesForMisplacedWeightsAndFunctions() throws IOEx
             queryParser.parse(query).query();
             fail("Expect exception here because boost_factor must not have a weight");
         } catch (QueryParsingException e) {
+            e.printStackTrace();
```
leftover?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/BaseQueryParserTemp.java:
```diff
@@ -33,7 +33,7 @@
     @Override
     public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOException, QueryParsingException {
-        Query query = parse(parseContext);
+        Query query = parse(parseContext.context());
```
@javanna This is one of the places where the new parse context still needs to refer back to the enclosed context, since it is used in fromXContent here temporarily. The whole class will finally got however.

javanna replied:
yea thats fine, this whole class is marked //norelease , that makes sense

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/CommonTermsQueryParser.java:
```diff
@@ -87,9 +87,6 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                         analyzer = parser.text();
-                        if (parseContext.analysisService().analyzer(analyzer) == null) {
-                            throw new QueryParsingException(parseContext, "[common] analyzer [" + parser.text() + "] not found");
-                        }
```
@javanna had to move this check to the query builder, will throw QueryShardException there now.

javanna replied:
good

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/SimpleQueryStringBuilder.java:
```diff
@@ -260,26 +262,44 @@ public QueryValidationException validate() {
+                    resolvedFieldsAndWeights.put(fField, fieldsAndWeights.get(fField));
+                }
+            }
+        }
```
This block moved here from the parser, because we need access to mapperService()

javanna replied:
looks great

cbuescher commented:
@javanna I rebased and went with renaming the old QueryParseContext to QueryShardContext now, with similar exception naming. I went though the changes, commented where I had to adapt two queries we already touched. Apart from that I think the PercolatorQueriesRegistry needs some attantion later, also the IndexQueryParserService, but that can be done in subsequent PRs. Also the `isFilter` handling needs to move to `toQuery`, but I'd also prefer doing that in a separate PR (there's also reminders for that here: https://github.com/elastic/elasticsearch/issues/12022)

javanna commented on core/src/main/java/org/apache/lucene/queryparser/classic/MapperQueryParser.java:
```diff
@@ -85,14 +85,14 @@
     }
 
-    public MapperQueryParser(QueryParserSettings settings, QueryParseContext parseContext) {
+    public MapperQueryParser(QueryParserSettings settings, QueryShardContext context) {
```
this constructor is unused, both in our branch and in master. We can remove it I think

javanna commented on core/src/main/java/org/elasticsearch/action/admin/indices/validate/query/TransportValidateQueryAction.java:
```diff
@@ -189,7 +190,10 @@ protected ShardValidateQueryResponse shardOperation(ShardValidateQueryRequest re
+        } catch (QueryShardException e) {
+            valid = false;
+            error = e.getDetailedMessage();
         } catch (QueryParsingException e) {
```
you can remove the QueryParsingException catch, it's unreachable

cbuescher replied:
Why? As far as I can see queryParserService.parseQuery(request.source()) can still throw QueryParsingException, I didn't change that exception there because that method itself does some xContent parsing.

javanna replied:
no, the catch there should be removed as well....that is what my IDE tells me :)

javanna replied:
ok this was my bad...I misread the warning :) just replace the two catches with a single `catch(QueryShardException|QueryParsingException e)` ?

javanna commented on core/src/main/java/org/elasticsearch/index/query/BaseQueryParser.java:
```diff
@@ -26,14 +26,14 @@
     @Override
-    public final Query parse(QueryParseContext parseContext) throws IOException, QueryParsingException {
-        return fromXContent(parseContext).toQuery(parseContext);
+    public final Query parse(QueryShardContext context) throws IOException, QueryParsingException {
```
I think conceptually this should be QueryParseContext instead, if it needs to do more  (toQuery) then we need to figure out how to create the QueryShardContext too out of it, but the other way around seems confusing to me. Sorry I see we are going back on forth on this.

javanna replied:
nevermind, I guess it depends on how you look at it. at the end of the day this parse method does parse + toQuery, having QueryShardContext is fine given that it will still happen on the shard. Also given that the QueryParseContext is much more lightweight, it makes more sense if done this way.

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -177,15 +181,17 @@ public ParsedQuery parse(QueryParseContext context, BytesReference source) {
     }
 
-    public ParsedQuery parse(String source) throws QueryParsingException {
+    public ParsedQuery parse(String source) throws QueryShardException, QueryShardException {
```
something odd happened here :)

javanna replied:
should we mark this //norelease ?

cbuescher replied:
The whole class maybe? There are so many parse methods here that do the whole roundtrip.

javanna replied:
I know, still the class will stay but these methods should go. I would mark every single method //norelease

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -177,15 +181,17 @@ public ParsedQuery parse(QueryParseContext context, BytesReference source) {
             return innerParse(cache.get(), parser);
+        } catch (QueryShardException e) {
+            throw e;
         } catch (QueryParsingException e) {
```
we can remove this catch

cbuescher replied:
After I removed this, some tests in SimpleIndexQueryParsersTest failed, they seem to make assertions in the exception message, so I leave it for now.

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -276,33 +277,39 @@ public ParsedQuery parseQuery(BytesReference source) {
             }
+        } catch (QueryShardException e) {
+            throw e;
         } catch (QueryParsingException e) {
```
I think this catch can go away

javanna replied:
again, I misread my IDE warning, just replace with a single catch

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -276,33 +277,39 @@ public ParsedQuery parseQuery(BytesReference source) {
 
-    private ParsedQuery innerParse(QueryParseContext parseContext, XContentParser parser) throws IOException, QueryParsingException {
-        parseContext.reset(parser);
+    private static ParsedQuery innerParse(QueryShardContext context, QueryBuilder queryBuilder) throws IOException, QueryShardException {
```
maybe move it after the method that calls it for readability?

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -224,27 +230,22 @@ public ParsedQuery parseInnerFilter(XContentParser parser) throws IOException {
-    public Query parseInnerQuery(QueryParseContext parseContext) throws IOException {
-        parseContext.parseFieldMatcher(parseFieldMatcher);
-        Query query = parseContext.parseInnerQuery();
+    public Query parseInnerQuery(QueryShardContext context) throws IOException {
```
should we mark this //norelease?

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -276,33 +277,39 @@ public ParsedQuery parseQuery(BytesReference source) {
+        return new ParsedQuery(query, context.copyNamedQueries());
+    }
+
+    private ParsedQuery innerParse(QueryShardContext context, XContentParser parser) throws IOException, QueryShardException {
```
should we mark this //norelease?

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -210,10 +216,10 @@ public ParsedQuery parse(QueryParseContext context, XContentParser parser) {
      */
     @Nullable
     public ParsedQuery parseInnerFilter(XContentParser parser) throws IOException {
```
should we mark this //norelease ?

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -197,11 +203,11 @@ public ParsedQuery parse(XContentParser parser) {
     }
 
-    public ParsedQuery parse(QueryParseContext context, XContentParser parser) {
+    public ParsedQuery parse(QueryShardContext context, XContentParser parser) {
```
should we mark this //norelease

javanna commented on core/src/main/java/org/elasticsearch/index/query/IndexQueryParserService.java:
```diff
@@ -161,7 +165,7 @@ public ParsedQuery parse(BytesReference source) {
     }
 
-    public ParsedQuery parse(QueryParseContext context, BytesReference source) {
+    public ParsedQuery parse(QueryShardContext context, BytesReference source) {
```
should we mark this //norelease ?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,107 @@
-        this.parser = parser;
+    //norelease this is still used in BaseQueryParserTemp and FunctionScoreQueryParse, remove if not needed there anymore
+    @Deprecated
+    public QueryShardContext context() {
```
can we rename this method?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,107 @@
+     */
+    @Nullable
+    @Deprecated
+    public Query parseInnerQuery() throws IOException, QueryShardException {
```
//norelease ?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -0,0 +1,322 @@
+//     */
+//    public boolean isFilter() {
+//        return isFilter;
+//    }
```
we have to restore this right?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardException.java:
```diff
@@ -0,0 +1,74 @@
+
+/**
+ *
+ */
```
javadocs please ;)

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -0,0 +1,322 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class QueryShardContext {
```
javadocs please ;)

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardException.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class QueryShardException extends ElasticsearchException {
+
+    static final int UNKNOWN_POSITION = -1;
```
this can go away?

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -312,12 +312,20 @@ public void testSerialization() throws IOException {
     /**
+     * @return a new {@link QueryShardContext} based on the base test index and queryParserService
+     */
+    protected static QueryShardContext createContext() {
```
rename the method to createShardContext ?

javanna commented:
I did another round, I think this is getting very close. The only issue I have is the fact that you can retrieve the QueryParseContext from the QueryShardContext. This confuses me, I find it conceptually wrong as you are in parse phase and might need to move on, hence create a QueryShardContext given a QueryParseContext. The other way around just because the latter is a lighter context compared to the former isn't a goo engough reason to do things this way to be honest, unless I am missing other reasons why we moved to this approach. Let me know what you think. I would still have all the parse methods accept a QueryParseContext and remove getParseContext from QueryShardContext.

cbuescher commented:
@javanna addressed most of your last comment, I have trouble removing catching those exceptions you mentioned, in one case test fails, in other case I still don't see why this should happen. Also need to have a final sync on the direction in which the two contexts refer to each other for now.

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,108 @@
-    private boolean allowUnmappedFields;
-
-    private boolean mapUnmappedFieldAsString;
+    private final QueryShardContext context;
```
can you add a //norelease here too? context should really go away after all queries are refactored

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,108 @@
+    //norelease should be possible to remove after refactoring all queries
+    public Query parseInnerFilter(String queryName) throws IOException, QueryShardException {
+        QueryBuilder builder = parseInnerFilterToQueryBuilder(queryName);
+        return (builder != null) ? builder.toQuery(this.context) : null;
```
I would add an `assert this.context != null` here just to make sure

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -19,201 +19,108 @@
+        QueryBuilder builder = parseInnerQueryBuilder();
+        Query result = null;
+        if (builder != null) {
+            result = builder.toQuery(this.context);
```
I would add an `assert this.context != null` here just to make sure

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -314,123 +193,19 @@ public QueryBuilder parseInnerFilterToQueryBuilder(String queryName) throws IOEx
 
-    public Version indexVersionCreated() {
-        return indexVersionCreated;
+    public QueryParser queryParser(String name) {
```
this doesn't need to be public right? package private maybe?

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java:
```diff
@@ -257,40 +164,12 @@ public QueryBuilder parseInnerQueryBuilder() throws IOException {
-    /**
      * @return
-     * @throws QueryParsingException
+     * @throws QueryShardException
```
can you fix the javadocs here while you are at it? it seems missing

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryShardException.java:
```diff
@@ -0,0 +1,72 @@
+        super.writeTo(out);
+    }
+
+    public QueryShardException(StreamInput in) throws IOException{
```
can we move this constructor on top?

javanna commented:
I did a final round, left a few minor comments, but this LGTM once those are addressed

cbuescher commented:
@javanna did another rebase, will still need to do a final squash and work on the commit message if you think this should go into the branch now
